src/services/review/ReviewRequestService.ts

```diff
@@ -196,13 +215,103 @@ export default class ReviewRequestService {
           createdAt: new Date(created_at).getTime(),
           // TODO!
           newComments: 0,
-          // TODO!
-          firstView: false,
+          firstView: isFirstView,
         }
       })
     )
   }
 
+  markAllReviewRequestsAsViewed = async (
+    sessionData: UserWithSiteSessionData,
+    site: Site
+  ): Promise<void> => {
+    const { isomerUserId: userId } = sessionData
+
+    const requestsViewed = await this.reviewRequestView.findAll({
+      where: {
+        siteId: site.id,
+        userId,
+      },
+    })
+
+    const allActiveRequests = await this.repository.findAll({
+      where: {
+        siteId: site.id,
+        // NOTE: Closed and merged review requests would not have an
+        // entry in the review request views table
+        reviewStatus: ["OPEN", "APPROVED"],
+      },
+    })
+
+    const requestIdsViewed = requestsViewed.map(
+      (request) => request.reviewRequestId
+    )
+    const allActiveRequestIds = allActiveRequests.map((request) => request.id)
+    const requestIdsToMarkAsViewed = _.difference(
+      allActiveRequestIds,
+      requestIdsViewed
+    )
+
+    requestIdsToMarkAsViewed.forEach(async (requestId) => {
+      await this.reviewRequestView.create({
+        reviewRequestId: requestId,
+        siteId: site.id,
+        userId,
+        // This field represents the user opening the review request
+        // itself, which the user has not done so yet at this stage.
+        lastViewedAt: null,
+      })
+    })
+  }
+
+  updateReviewRequestLastViewedAt = async (
+    sessionData: UserWithSiteSessionData,
+    site: Site,
+    pullRequestNumber: number
+  ): Promise<void | RequestNotFoundError> => {
+    const { isomerUserId: userId } = sessionData
+    const reviewRequest = await this.getReviewRequest(site, pullRequestNumber)
+
+    if (isIsomerError(reviewRequest)) {
+      return reviewRequest
+    }
+
+    const { id: reviewRequestId } = reviewRequest
+
+    await this.reviewRequestView.update(
+      {
+        lastViewedAt: new Date(),
+      },
+      {
+        where: {
+          reviewRequestId,
+          siteId: site.id,
+          userId,
+        },
+      }
+    )
+  }
+
+  deleteAllReviewRequestViews = async (
```

Comment:
just curious, why do we need to delete the views? is this to avoid paying storage costs?

Reply:
Yup that's correct. Plus it is in case we want to allow reopening of closed RRs, then we want to show it as a new RR for everyone.